bigframes/session/executor.py

```diff
@@ -0,0 +1,329 @@
+# Copyright 2024 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+import math
+from typing import cast, Iterable, Literal, Mapping, Optional, Sequence, Tuple
+import warnings
+import weakref
+
+import google.api_core.exceptions
+import google.cloud.bigquery as bigquery
+import google.cloud.bigquery.job as bq_job
+
+import bigframes.core
+import bigframes.core.compile
+import bigframes.core.guid
+import bigframes.core.nodes as nodes
+import bigframes.core.ordering as order
+import bigframes.core.tree_properties as tree_properties
+import bigframes.formatting_helpers as formatting_helpers
+import bigframes.session._io.bigquery as bq_io
+import bigframes.session.metrics
+import bigframes.session.planner
+import bigframes.session.temp_storage
+
+# Max complexity that should be executed as a single query
+QUERY_COMPLEXITY_LIMIT = 1e7
+# Number of times to factor out subqueries before giving up.
+MAX_SUBTREE_FACTORINGS = 5
+
+_MAX_CLUSTER_COLUMNS = 4
+
+
+class BigQueryCachingExecutor:
```

Comment:
Could you please add some comments to briefly explain how the cached executions works.

Reply:
Added a bit of an explantation to class description.